module/Finna/src/Finna/Service/RemsService.php

```diff
@@ -0,0 +1,826 @@
+            break;
+        case 'session-expired':
+            $status = self::STATUS_EXPIRED;
+            break;
+        default:
+            $status = self::STATUS_CLOSED;
+        }
+        $this->session->{self::SESSION_ACCESS_STATUS} = $status;
+    }
+
+    /**
+     * Set blacklist status of current user.
+     * This is called from R2 backend connector.
+     *
+     * @param string|null $status Blacklist added date or
+     * null if the user is not blacklisted.
+     *
+     * @return void
+     */
+    public function setBlacklistStatusFromConnector($status)
+    {
+        $this->session->{self::SESSION_BLACKLISTED} = $status;
+    }
+
+    /**
+     * Prepare user id for saving to REMS.
+     *
+     * @param string $userId User id.
+     *
+     * @return string
+     */
+    public function prepareUserId($userId)
+    {
+        // Strip organisation-id (subdomain) from username
+        if (false !== strpos($userId, ':')) {
+            list($domain, $userId) = explode(':', $userId, 2);
+        }
+        return $userId;
+    }
+
+    /**
+     * Get application.
+     *
+     * @param int  $id    Id
+     * @param bool $throw Throw exception?
+     *
+     * @throws Exception
+     * @return array|null
+     */
+    protected function getApplication($id, $throw = false)
+    {
+        return $this->sendRequest(
+            "applications/$id", [], 'GET', RemsService::TYPE_USER,
+            null, false, $throw
+        );
+    }
+
+    /**
+     * Get user application ids.
+     *
+     * @param array $statuses application statuses
+     *
+     * @return array
+     */
+    protected function getApplications($statuses = [])
+    {
+        // Fetching applications by query doesn't work with REMS api.
+        // Therefore fetch all and filter by status manually.
+
+        $params = [];
+        try {
+            $result = $this->sendRequest(
+                'my-applications',
+                $params, 'GET', RemsService::TYPE_USER, null, false
+            );
+        } catch (\Exception $e) {
+            return [];
+        }
+
+        if ($statuses) {
+            $statuses = array_map(
+                function ($status) {
+                    return "application.state/$status";
+                },
+                $statuses
+            );
+
+            $result = array_filter(
+                $result,
+                function ($application) use ($statuses) {
+                    return in_array($application['application/state'], $statuses);
+                }
+            );
+        }
+
+        return array_map(
+            function ($application) {
+                return ['id' => $application['application/id']];
+            },
+            $result
+        );
+    }
+
+    /**
+     * Return REMS user id (eppn) of the current authenticated user.
+     *
+     * @return string
+     * @throws Exception
+     */
+    protected function getUserId()
+    {
+        if (!$user = $this->auth->isLoggedIn()) {
```

Comment:
$this->auth ei käytetä muuhun kuin tähän, laita käyttäjätunnus suoraan constructoriin ennemmin.